### PR TITLE
Technical refactoring permissions

### PR DESCRIPTION
**Pull request by Fancien**
- Refactoring permission system to include permission target in the permission codename, increasing code readability
- Adding permission requirements to model views
- Adding tests to assert that urls return a 403 html error when user does not have permission to access that url.
- On dataset permissions management page, added the list of users with permissions inherited from parent project
- On user management page, adding the list of entities on which this user has permission
- Forbid users to change Access status to "Active" if user is not data steward

### CODE REVIEW

File: core/permissions/checker.py [hunk 1]
```diff
@@ -141,39 +199,40 @@ def _check(self, perm, obj, **kwargs):
...
-        else:
-            return False
+        parent_checker = DatasetChecker(self.user_or_group, checker=self.checker)
+        return parent_checker.check(perm, parent_dataset)
 
 
 class AutoChecker(AbstractChecker):
     """
-    Check permission on given object. Object must be instance on a key in the __mapping attribute.
+    Check permission on given object. Object must be instanced on a key in the __mapping attribute.
```
Comment (neoflex) [inline core/permissions/checker.py:223]: sentence not very clear to me

File: web/templates/datasets/dataset.html [hunk 2]
```diff
@@ -57,25 +57,26 @@ <h1>{{ dataset.title }}</h1>
...
+                        </a>
+                    {% elif request.user.can_publish %}
+                        <a href="{% url 'dataset_unpublish' pk=dataset.pk %}"
+                           data-confirm="Are you sure to unpublish the dataset? The changes won't be propagated to other systems!"
+                           id="unpublish"
+                           title="Unpublish dataset"
+                           class="btn btn-default bmd-btn-fab float-right">
+                            <i class="material-icons">undo</i>
+                        </a>
+                    {% endif %}
```
Comment (neoflex) [inline web/templates/datasets/dataset.html:79]: IMO this king of branching logic is easier to read with nested if:

```
{% if request.user.can_publish %}
      {% if dataset.is_published %}
...
      {% else %}
...
{% endif %}
```

File: web/templates/permissions.html [hunk 3]
```diff
@@ -28,37 +28,47 @@ <h2 class="card-title">Local custodians</h2>
...
+                <h2 class="card-title">View Inherited Permissions from project <a href="{% url 'project' pk=object.project.id %}">{{ object.project }}</a></h2>
+                <div class="card-text md-1">
+                    {% include '_includes/permission_table.html' with obj_name='User' obj_perms_const=pj_perms_const obj_perms=inherited_permissions %}
+                </div>
+            </div>
+            <div class="card-footer">
+                You can change the permissions on the <a href="{% url 'permission_project' pk=object.project.id %}">project permission edit page</a>.
+            </div>
+        </div>
+    </div>
```
Comment (neoflex) [inline web/templates/permissions.html:51]: indentation

**Review by neoflex — CHANGES_REQUESTED** [review]

File: core/models/access.py [hunk 4]
```diff
@@ -138,10 +139,7 @@ def clean(self):
     )
 
     def __str__(self):
-        if self.contact:
-            return f"Access ({self.status}) to dataset {self.dataset.title} given to a user: {self.user}/{self.access_notes}"
-        else:
-            return f"Access ({self.status}) to dataset {self.dataset.title} given to a contact: {self.user}/{self.access_notes}"
+        return f"Access ({self.status}) to dataset {self.dataset.title} given to a contact: {self.user}/{self.access_notes}"
```
[the remark this answers is not in the document]
Comment (Fancien) [reply core/models/access.py:142]: The issue is that the `if self.contact` caused an error when the contact is deleted. Due to the cascade deletion, the access is also deleted.

A LogEntry is then created to log the deletion of the access, but then the `self.contact` in this method returns a deleted object, which raises an error.

File: README.md [hunk 5]
```diff
@@ -238,7 +238,7 @@ npm run-script build
 ### Run the built-in web server (for development)
 
 ```bash
-./manage.py runserver
+./manage.py "runserver
```
[the remark this answers is not in the document]
Comment (Fancien) [reply README.md]: This is corrected. I removed the `load_initial_data` section in upgrade procedure and corrected the typo

File: core/models/access.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Fancien) [reply core/models/access.py]: This should be corrected. I added a try/except block to deal with the deletion of contacts

**Pull request merged**